File: typescript/src/services/projects.ts
import {
  Project,
  GraphQLRequest,
  ProjectsResponse,
} from '../types/projects';
import { AuthService } from './auth';
import { UserService } from './user';
import { IFileSystem } from '../types/filesystem';

/**
 * Projects service for managing JuliaHub projects
 * Migrated from projects.go and git.go (project lookup parts)
 */
export class ProjectsService {
  private authService: AuthService;
  private userService: UserService;

  constructor(fs: IFileSystem) {
    this.authService = new AuthService(fs);
    this.userService = new UserService(fs);
  }

  /**
   * Execute a GraphQL projects query
   */
  private async executeProjectsQuery(
    server: string,
    query: string,
    userID: number
  ): Promise<ProjectsResponse> {
    const token = await this.authService.ensureValidToken();

    const graphqlReq: GraphQLRequest = {
      operationName: 'Projects',
      query: query,
      variables: {
        ownerId: userID,
      },
    };

    const url = `https://${server}/v1/graphql`;
    const resp = await fetch(url, {
      method: 'POST',
      headers: {
        Authorization: `Bearer ${token.idToken}`,
        'Content-Type': 'application/json',
        Accept: 'application/json',
        'X-Hasura-Role': 'jhuser',
      },
      body: JSON.stringify(graphqlReq),
    });

    if (!resp.ok) {
      const errorText = await resp.text();
      throw new Error(
        `GraphQL request failed (status ${resp.status}): ${errorText}`
      );
    }

    const response = (await resp.json()) as ProjectsResponse;

    // Check for GraphQL errors
    if (response.errors && response.errors.length > 0) {
      throw new Error(`GraphQL errors: ${JSON.stringify(response.errors)}`);
    }

    return response;
  }

  /**
   * List all projects with optional user filtering
   */
  async listProjects(
    server: string,
    userFilter?: string,
    userFilterProvided: boolean = false
  ): Promise<string> {
    // Get user info to get the user ID
    const userInfo = await this.userService.getUserInfo(server);

    // GraphQL query
    const query = `query Projects(
  $limit: Int
    $offset: Int
      $orderBy: [projects_order_by!]
        $ownerId: bigint
          $filter: projects_bool_exp
            ) {
              aggregate: projects_aggregate(where: $filter) {
                aggregate {
                  count
                }
              }
              projects(limit: $limit, offset: $offset, order_by: $orderBy, where: $filter) {
                id: project_id
                project_id
                name
                owner {
                  username
                  name
                }
                created_at
                product_id
                finished
                is_archived
                instance_default_role
                deployable
                project_deployments_aggregate {
                  aggregate {
                    count
                  }
                }
                running_deployments: project_deployments_aggregate(
                  where: {
                    status: { _eq: "JobQueued" }
                    job: { status: { _eq: "Running" } }
                  }
                ) {
                  aggregate {
                    count
                  }
                }
                pending_deployments: project_deployments_aggregate(
                  where: {
                    status: { _eq: "JobQueued" }
                    job: { status: { _in: ["SubmitInitialized", "Submitted", "Pending"] } }
                  }
                ) {
                  aggregate {
                    count
                  }
                }
                resources(order_by: [{ sorting_order: asc_nulls_last }]) {
                  sorting_order
                  instance_default_role
                  giturl
                  name
                  resource_id
                  resource_type
                }
                product {
                  id
                  displayName: display_name
                  name
                }
                visibility
                description
                users: groups(where: { group_id: { _is_null: true } }) {
                  user {
                    name
                  }
                  id
                  assigned_role
                }
                groups(where: { group_id: { _is_null: false } }) {
                  group {
                    name
                    group_id
                  }
                  id: group_id
                  group_id
                  project_id
                  assigned_role
                }
                tags
                userRole: access_control_users_aggregate(
                  where: { user_id: { _eq: $ownerId } }
                ) {
                  aggregate {
                    max {
                      assigned_role
                    }
                  }
                }
                is_simple_mode
                projects_current_editor_user_id {
                  name
                  id
                }
              }
            }`;

    const response = await this.executeProjectsQuery(server, query, userInfo.id);
    let projects = response.data.projects;

    // Apply user filtering if requested
    if (userFilterProvided) {
      if (!userFilter) {
        // Show only current user's projects
        projects = projects.filter((p) => p.owner.username === userInfo.username);
      } else {
        // Show projects from specified user
        projects = projects.filter(
          (p) => p.owner.username.toLowerCase() === userFilter.toLowerCase()
        );
      }
    }

    if (projects.length === 0) {
      if (userFilterProvided) {
        if (!userFilter) {
          return 'No projects found for your user';
        } else {
          return `No projects found for user '${userFilter}'`;
        }
      } else {
        return 'No projects found';
      }
    }

    let output = '';
    if (userFilterProvided) {
      if (!userFilter) {
        output += `Found ${projects.length} project(s) for your user:\n\n`;
      } else {
        output += `Found ${projects.length} project(s) for user '${userFilter}':\n\n`;
      }
    } else {
      output += `Found ${projects.length} project(s):\n\n`;
    }

    for (const project of projects) {
      output += this.formatProject(project);
      output += '\n';
    }

    return output;
  }

  /**
   * Format a single project for display
   */
  private formatProject(project: Project): string {
    let output = '';
    output += `ID: ${project.id}\n`;
    output += `Name: ${project.name}\n`;
    output += `Owner: ${project.owner.username} (${project.owner.name})\n`;

    if (project.description) {
      output += `Description: ${project.description}\n`;
    }

    output += `Visibility: ${project.visibility}\n`;
    output += `Product: ${project.product.displayName}\n`;
    output += `Created: ${project.created_at}\n`;
    output += `Finished: ${project.finished}\n`;
    output += `Archived: ${project.is_archived}\n`;
    output += `Deployable: ${project.deployable}\n`;

    // Show deployment counts
    const totalDeployments = project.project_deployments_aggregate.aggregate.count;
    const runningDeployments = project.running_deployments.aggregate.count;
    const pendingDeployments = project.pending_deployments.aggregate.count;
    output += `Deployments: ${totalDeployments} total, ${runningDeployments} running, ${pendingDeployments} pending\n`;

    // Show resources
    if (project.resources.length > 0) {
      output += 'Resources:\n';
      for (const resource of project.resources) {
        output += `  - ${resource.name} (${resource.resource_type})\n`;
        if (resource.giturl) {
          output += `    Git URL: ${resource.giturl}\n`;
        }
      }
    }

    // Show tags
    if (project.tags.length > 0) {
      output += `Tags: ${project.tags.join(', ')}\n`;
    }

    // Show user role
    if (project.userRole.aggregate.max.assigned_role) {
      output += `Your Role: ${project.userRole.aggregate.max.assigned_role}\n`;
    }

    return output;
  }

  /**
   * Find a project by username and project name
   */
  async findProjectByUserAndName(
    server: string,
    username: string,
    projectName: string
  ): Promise<string> {
    // Get user info to get the user ID
    const userInfo = await this.userService.getUserInfo(server);

    // Use the same GraphQL query as listProjects
    const query = `query Projects(
  $limit: Int
    $offset: Int
      $orderBy: [projects_order_by!]
        $ownerId: bigint
          $filter: projects_bool_exp
            ) {
              aggregate: projects_aggregate(where: $filter) {
                aggregate {
                  count
                }
              }
              projects(limit: $limit, offset: $offset, order_by: $orderBy, where: $filter) {
                id: project_id
                project_id
                name
                owner {
                  username
                  name
                }
                created_at
                product_id
                finished
                is_archived
                instance_default_role
                deployable
                project_deployments_aggregate {
                  aggregate {
                    count
                  }
                }
                running_deployments: project_deployments_aggregate(
                  where: {
                    status: { _eq: "JobQueued" }
                    job: { status: { _eq: "Running" } }
                  }
                ) {
                  aggregate {
                    count
                  }
                }
                pending_deployments: project_deployments_aggregate(
                  where: {
                    status: { _eq: "JobQueued" }
                    job: { status: { _in: ["SubmitInitialized", "Submitted", "Pending"] } }
                  }
                ) {
                  aggregate {
                    count
                  }
                }
                resources(order_by: [{ sorting_order: asc_nulls_last }]) {
                  sorting_order
                  instance_default_role
                  giturl
                  name
                  resource_id
                  resource_type
                }
                product {
                  id
                  displayName: display_name
                  name
                }
                visibility
                description
                users: groups(where: { group_id: { _is_null: true } }) {
                  user {
                    name
                  }
                  id
                  assigned_role
                }
                groups(where: { group_id: { _is_null: false } }) {
                  group {
                    name
                    group_id
                  }
                  id: group_id
                  group_id
                  project_id
                  assigned_role
                }
                tags
                userRole: access_control_users_aggregate(
                  where: { user_id: { _eq: $ownerId } }
                ) {
                  aggregate {
                    max {
                      assigned_role
                    }
                  }
                }
                is_simple_mode
                projects_current_editor_user_id {
                  name
                  id
                }
              }
            }`;

    const response = await this.executeProjectsQuery(server, query, userInfo.id);

    // Search for the project
    const matchedProject = response.data.projects.find(
      (project) =>
        project.owner.username.toLowerCase() === username.toLowerCase() &&
        project.name.toLowerCase() === projectName.toLowerCase()
    );

    if (!matchedProject) {
      throw new Error(`Project '${projectName}' not found for user '${username}'`);
    }

    console.log(
      `Found project: ${matchedProject.name} by ${matchedProject.owner.username} (ID: ${matchedProject.id})`
    );
    return matchedProject.id;
  }
}

File: projects.go
package main

import (
	_ "embed"
	"encoding/json"
	"fmt"
	"strings"
)

//go:embed projects.gql
var projectsQuery string

type Project struct {
	ID                          string       `json:"id"`
	ProjectID                   string       `json:"project_id"`
	Name                        string       `json:"name"`
	Owner                       ProjectOwner `json:"owner"`
	CreatedAt                   string       `json:"created_at"`
	ProductID                   int64        `json:"product_id"`
	Finished                    bool         `json:"finished"`
	IsArchived                  bool         `json:"is_archived"`
	InstanceDefaultRole         string       `json:"instance_default_role"`
	Deployable                  bool         `json:"deployable"`
	ProjectDeploymentsAggregate struct {
		Aggregate struct {
			Count int `json:"count"`
		} `json:"aggregate"`
	} `json:"project_deployments_aggregate"`
	RunningDeployments struct {
		Aggregate struct {
			Count int `json:"count"`
		} `json:"aggregate"`
	} `json:"running_deployments"`
	PendingDeployments struct {
		Aggregate struct {
			Count int `json:"count"`
		} `json:"aggregate"`
	} `json:"pending_deployments"`
	Product     Product  `json:"product"`
	Visibility  string   `json:"visibility"`
	Description string   `json:"description"`
	Tags        []string `json:"tags"`
	UserRole    struct {
		Aggregate struct {
			Max struct {
				AssignedRole string `json:"assigned_role"`
			} `json:"max"`
		} `json:"aggregate"`
	} `json:"userRole"`
	IsSimpleMode                bool `json:"is_simple_mode"`
	ProjectsCurrentEditorUserID struct {
		Name string `json:"name"`
		ID   int64  `json:"id"`
	} `json:"projects_current_editor_user_id"`
}

type ProjectOwner struct {
	Username string `json:"username"`
	Name     string `json:"name"`
}

type Product struct {
	ID          int64  `json:"id"`
	DisplayName string `json:"displayName"`
	Name        string `json:"name"`
}

type GraphQLRequest struct {
	OperationName string                 `json:"operationName"`
	Query         string                 `json:"query"`
	Variables     map[string]interface{} `json:"variables"`
}

type ProjectsResponse struct {
	Data struct {
		Projects  []Project `json:"projects"`
		Aggregate struct {
			Aggregate struct {
				Count int `json:"count"`
			} `json:"aggregate"`
		} `json:"aggregate"`
	} `json:"data"`
	Errors []struct {
		Message string `json:"message"`
	} `json:"errors"`
}

// projectsPageSize is how many projects are fetched per GraphQL request.
// The Projects query is expensive per row on the server (several aggregates
// and permission checks per project), so it is always paginated and filtered
// server-side; never ask Hasura for the whole projects table.
const projectsPageSize = 100

// escapeLikePattern escapes the LIKE/ILIKE metacharacters in s so that it
// matches literally when used as an _ilike pattern.
func escapeLikePattern(s string) string {
	s = strings.ReplaceAll(s, `\`, `\\`)
	s = strings.ReplaceAll(s, `%`, `\%`)
	s = strings.ReplaceAll(s, `_`, `\_`)
	return s
}

// buildProjectsFilter returns the projects_bool_exp to send to Hasura.
// Filtering happens server-side so the caller only receives (and the server
// only computes) the projects that will actually be shown.
func buildProjectsFilter(userFilter string, userFilterProvided bool, currentUserID int64) map[string]interface{} {
	if !userFilterProvided {
		return map[string]interface{}{}
	}
	if userFilter == "" {
		return map[string]interface{}{
			"owner_id": map[string]interface{}{"_eq": currentUserID},
		}
	}
	// Case-insensitive exact match on the owner's username (was EqualFold client-side).
	return map[string]interface{}{
		"owner": map[string]interface{}{
			"username": map[string]interface{}{"_ilike": escapeLikePattern(userFilter)},
		},
	}
}

// fetchProjectsPage requests one page of projects matching filter. It returns
// the page's projects and the total number of matching projects on the server.
func fetchProjectsPage(server string, token *StoredToken, ownerID int64, filter map[string]interface{}, page int) ([]Project, int, error) {
	body, err := executeGraphQL(server, token, GraphQLRequest{
		OperationName: "Projects",
		Query:         projectsQuery,
		Variables: map[string]interface{}{
			"ownerId": ownerID,
			"filter":  filter,
			"limit":   projectsPageSize,
			"offset":  (page - 1) * projectsPageSize,
			"orderBy": map[string]interface{}{"created_at": "desc_nulls_last"},
		},
	})
	if err != nil {
		return nil, 0, err
	}

	var response ProjectsResponse
	if err := json.Unmarshal(body, &response); err != nil {
		return nil, 0, fmt.Errorf("failed to parse response: %w", err)
	}
	if len(response.Errors) > 0 {
		return nil, 0, fmt.Errorf("GraphQL errors: %v", response.Errors)
	}
	return response.Data.Projects, response.Data.Aggregate.Aggregate.Count, nil
}

// totalPages is how many pages of projectsPageSize the given total spans.
func totalPages(total int) int {
	if total <= 0 {
		return 0
	}
	return (total + projectsPageSize - 1) / projectsPageSize
}

func listProjects(server string, userFilter string, userFilterProvided bool, page int) error {
	if page < 1 {
		return fmt.Errorf("--page must be 1 or greater, got %d", page)
	}

	token, err := ensureValidToken()
	if err != nil {
		return fmt.Errorf("authentication required: %w", err)
	}

	// Get user info to get the user ID
	userInfo, err := getUserInfo(server)
	if err != nil {
		return fmt.Errorf("failed to get user info: %w", err)
	}

	filter := buildProjectsFilter(userFilter, userFilterProvided, userInfo.ID)
	projects, total, err := fetchProjectsPage(server, token, userInfo.ID, filter, page)
	if err != nil {
		return err
	}

	if total == 0 {
		if userFilterProvided {
			if userFilter == "" {
				fmt.Println("No projects found for your user")
			} else {
				fmt.Printf("No projects found for user '%s'\n", userFilter)
			}
		} else {
			fmt.Println("No projects found")
		}
		return nil
	}

	pages := totalPages(total)
	if len(projects) == 0 {
		fmt.Printf("Page %d is empty: %d project(s) found, %d page(s) of %d\n", page, total, pages, projectsPageSize)
		return nil
	}

	if userFilterProvided {
		if userFilter == "" {
			fmt.Printf("Found %d project(s) for your user:\n\n", total)
		} else {
			fmt.Printf("Found %d project(s) for user '%s':\n\n", total, userFilter)
		}
	} else {
		fmt.Printf("Found %d project(s):\n\n", total)
	}
	if pages > 1 {
		fmt.Printf("Showing page %d of %d (%d per page, most recently created first; use --page to see more)\n\n", page, pages, projectsPageSize)
	}

	for _, project := range projects {
		fmt.Printf("ID: %s\n", project.ID)
		fmt.Printf("Name: %s\n", project.Name)
		fmt.Printf("Owner: %s (%s)\n", project.Owner.Username, project.Owner.Name)
		if project.Description != "" {
			fmt.Printf("Description: %s\n", project.Description)
		}
		fmt.Printf("Visibility: %s\n", project.Visibility)
		fmt.Printf("Product: %s\n", project.Product.DisplayName)
		fmt.Printf("Created: %s\n", project.CreatedAt)
		fmt.Printf("Finished: %t\n", project.Finished)
		fmt.Printf("Archived: %t\n", project.IsArchived)
		fmt.Printf("Deployable: %t\n", project.Deployable)

		// Show deployment counts
		totalDeployments := project.ProjectDeploymentsAggregate.Aggregate.Count
		runningDeployments := project.RunningDeployments.Aggregate.Count
		pendingDeployments := project.PendingDeployments.Aggregate.Count
		fmt.Printf("Deployments: %d total, %d running, %d pending\n",
			totalDeployments, runningDeployments, pendingDeployments)

		// Show tags
		if len(project.Tags) > 0 {
			fmt.Printf("Tags: %v\n", project.Tags)
		}

		// Show user role
		if project.UserRole.Aggregate.Max.AssignedRole != "" {
			fmt.Printf("Your Role: %s\n", project.UserRole.Aggregate.Max.AssignedRole)
		}

		fmt.Println()
	}

	return nil
}
